src/QirRuntime/public/QuantumApiBase.hpp

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#include <exception>
+
+#include "IQuantumApi.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    class CQuantumApiBase : public IQuantumApi
+    {
+      public:
+        // Qubit management
+        Qubit AllocateQubit() override
+        {
+            throw std::logic_error("not_implemented");

```

Comment:
Is the expectation that they will change to not throw sometime soon? If these are essentially pure virtual methods in an abstract base class, then why not have them be pure virtual methods in an abstract base class?

Throwing a "not implemented" exception on a permanent basis means that you're claiming to implement some interface that you don't really implement -- you're lying to the compiler, either about your class (it doesn't really implement the interface) or the interface (it's much bigger than it should be) or both. See [here](https://www.dotnetperls.com/notimplementedexception) for a mild explanation, or [here](https://lostechies.com/jimmybogard/2008/10/12/notimplementedexception-and-the-interface-segregation-principle/) for more of a rant on the topic...

In this case, it seems that the only reason for this class to exist is to give the public virtual methods in IQuantumApi "not implemented" stub implementations. Assuming that any actual simulator will need to implement all of these, I think it would be better to just have the simulators inherit from IQuantumApi and leave methods "not implemented" in the actual simulator code during development, because then it's easy to see what work is remaining. If, instead, actual simulators don't need to actually implement all of these methods, then I would limit IQuantumAPI to only declare the methods that are actually necessary. Either way, I don't think this class is needed.

Reply:
I am well aware of the conversation around pros and cons of providing this kind of base classes. As you can see in this project, they are super useful for testing, where most test IQuantumAPIs only need to override/mock a few methods. If we collectively agree that none of the production code should use this base class, I will eliminate it from the corresponding implementations.

Reply:
I'm OK with it not being used in production code. What would be ideal would be a general IDE facility like VS has for C#, where you can start with an empty class that implements an interface and have VS generate the "not implemented" methods in the final class for you; I think that's rather out of scope for this PR, though :-)..